[WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation

lobstergrindset opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49547 [WIP] JIT Static Hooks: cpp tests
* #49546 [WIP] JIT Static Hooks: python tests
* #49545 [WIP] JIT Static Hooks: serialization logic
* #49975 [WIP] JIT Static Hooks: schema checking logic
* **#49544 [WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation**

Added support for module level hooks to be statically scripted: https://fb.quip.com/0IdqABRaPsAZ 
Allows hooks added with register_forward_hook and register_forward_pre_hook to be scripted and ran in python/cpp.

Major implementation changes:
- hooks/pre_hooks are compiled like methods on a module but are called with CallFunction
- hooks/pre_hooks of submodules have their calls injected before/after the submodule forward call in the IR
- hooks/pre_hooks of the main module are still called from python but are ran as script functions
- hooks/pre_hooks added to concrete_type and class_type 
- compilation unit compiles hooks (stores them as methods, hooks can't have the same name as a method on a module)

Differential Revision: [D25771122](https://our.internmc.facebook.com/intern/diff/D25771122)

Commit:
[WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation"




[ghstack-poisoned]

SplitInfinity reviewed (commented):
I think this is a good start but I have two high level concerns:
- It seems that hooks are allowed to be accessed on a module via attribute lookup as if they were methods; I don't think this should be allowed (unless it's possible in eager mode?).
- Hooks for `RecursiveScriptModule` seem to be registered as Python functions via `register_forward_{pre}_hook` instead of scripted functions

Lilyjjo commented on torch/jit/_recursive.py:
```diff
@@ -274,6 +274,14 @@ def infer_type(name, item):
             concrete_type_builder.add_failed_attribute(name, hint)
 
+    # add hooks to concrete type
+    # TODO: would using a random number here be better? I had collisions happen
+    # even with the name and python id combination here ...  
```
When I was writing the CPP testing code (test/jit_hooks/model.py in the top commit in this stack) I was running into hook_name + python_id collisions. Since python objects only own their id when they're in scope, the way that I was defining hooks inside of other functions was causing their python_ids to be re used. Specifically the test test_module_forward_and_pre_hook_single_IO, if you change the hook names to just be 'pre_hook' and 'forward_hook' it'll cause Python 3.8.5 to re-use the same python id with the same name pairing.

SplitInfinity replied:
Hm, since the goal here is to keep track of which hooks are attached to a module in order to figure out whether two modules have the same type, I wonder if you could store the actual function object in the concrete type (similar to what we do with function attributes) to get around this issue.

Lilyjjo replied:
Using the py::object works! It solves for the collision that was happening in the CPP testing code as well

Lilyjjo commented on torch/csrc/jit/python/python_sugared_value.h:
```diff
@@ -172,15 +172,97 @@ struct VISIBILITY_HIDDEN ModuleValue : public SugaredValue {
       size_t n_binders) override {
-    return attr(loc, caller, "forward")
-        ->call(loc, caller, args, kwargs, n_binders);
+    bool is_class_type =
+        concreteType_->getJitType()->cast<ClassType>() != nullptr;
+    bool have_pre_hooks = is_class_type &&
+        concreteType_->getJitType()
+                ->expect<ClassType>()
+                ->getForwardPreHooks()
+                .size() != 0;
+    bool have_hooks = is_class_type &&
+        concreteType_->getJitType()
+                ->expect<ClassType>()
+                ->getForwardHooks()
+                .size() != 0;
```
I was hitting the issue where InterfaceTypes were being passed as the self's type. The 'expect<ClassType>' call when trying to get the hooks was causing an internal assert to be thrown, I needed a guard against that. Does that justify the need for the is_class_type or am I misreading your comment?

Commit:
Update on "[WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation"


Implementation of design laid out in: https://fb.quip.com/MY9gAqlroo0Z

Major changes:
- hooks/pre_hooks are compiled and treated similarly to methods on a module
- hooks/pre_hooks of submodules have their calls injected before/after the submodule forward call in the IR
- hooks/pre_hooks of the main module are still called from python
- hooks/pre_hooks added to concrete_type and class_type 
- compilation unit compiles hooks (stores them as methods, hooks can't have the same name as a method on the module)

Work left to do:
- having hooks accept default arguments if forward has default arguments (untested)
- writing checks and error messages for wrong hook/pre_hook function signatures (this is a huge need, currently wrong signatures will compile and then have runtime errors or will segfault)
- debug seg fault on jit's legacy fuser

[ghstack-poisoned]

Lilyjjo commented:
The last commit update addressed the two larger issues that @SplitInfinity brought up and some of the NITs.
(1) Hooks are now not accessible on the module via attribute lookups
(2) the hooks on the script modules on the python side now run ScriptedFunction versions of the hooks
The only major change implementation wise is that now hooks are being called with CallFunction instead of CallMethod.

Still left to do: function schema error message checks and fine tuning the tuple argument treatment for submodule calls in module.cpp/python_sugared_value.h

Commit:
Update on "[WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation"


Implementation of design laid out in: https://fb.quip.com/MY9gAqlroo0Z

Major changes:
- hooks/pre_hooks are compiled and treated similarly to methods on a module
- hooks/pre_hooks of submodules have their calls injected before/after the submodule forward call in the IR
- hooks/pre_hooks of the main module are still called from python
- hooks/pre_hooks added to concrete_type and class_type 
- compilation unit compiles hooks (stores them as methods, hooks can't have the same name as a method on the module)

Work left to do:
- having hooks accept default arguments if forward has default arguments (untested)
- writing checks and error messages for wrong hook/pre_hook function signatures (this is a huge need, currently wrong signatures will compile and then have runtime errors or will segfault)
- debug seg fault on jit's legacy fuser

[ghstack-poisoned]

Lilyjjo commented on torch/csrc/jit/api/module.cpp:
```diff
@@ -129,6 +129,51 @@ c10::intrusive_ptr<c10::ivalue::Future> Method::run_async(
+
+  // call forward pre_hooks
+  for (const auto& pre_hook : pre_forward_hooks) {
+    std::vector<IValue> pre_hook_input_vals = inputs;
+    if (pre_hook_input_vals.size() == 0) {
+      pre_hook_input_vals.emplace_back(IValue());
+    }
+    auto tuple_input = c10::ivalue::Tuple::create(pre_hook_input_vals);
```
The inclusion of the inputs is to placate the typing issue described in the changes in pybind_utils.cpp. This error message gets thrown when the empty value isn't added:  "Expected a value of type 'Tuple[None]' for argument 'input' but instead found type 'Tuple[]'". I'm not sure how to get 'Tuple[]' to register as a valid type from the python side.

Lilyjjo replied:
We don't want to modify the inputs by adding the `None` value because the call to forward after the pre_hooks is expecting only the self argument

Lilyjjo commented on torch/csrc/jit/api/module.cpp:
```diff
@@ -129,6 +129,51 @@ c10::intrusive_ptr<c10::ivalue::Future> Method::run_async(
+    IValue result = Method(_ivalue(), pre_hook)(pre_hook_inputs);
+
+    if (!result.isNone()) {
+      inputs = result.toTuple()->elements();
```
This did cause errors in the cpp code when the return wasn't wrapped in a tuple inside of the hook, it's changed in the next commit

Commit:
Update on "[WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation"


Implementation of design laid out in: https://fb.quip.com/MY9gAqlroo0Z

Major changes:
- hooks/pre_hooks are compiled like methods on a module but are called with CallFunction
- hooks/pre_hooks of submodules have their calls injected before/after the submodule forward call in the IR
- hooks/pre_hooks of the main module are still called from python but are ran as script functions
- hooks/pre_hooks added to concrete_type and class_type 
- compilation unit compiles hooks (stores them as methods, hooks can't have the same name as a method on a module)

Work left to do:
- having hooks accept default arguments if forward has default arguments (untested)
- writing checks and error messages for wrong hook/pre_hook function signatures (this is a huge need, currently wrong signatures will compile and then have runtime errors or will segfault)
- debug seg fault on jit's legacy fuser

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation"


Implementation of design laid out in: https://fb.quip.com/MY9gAqlroo0Z

Major changes:
- hooks/pre_hooks are compiled like methods on a module but are called with CallFunction
- hooks/pre_hooks of submodules have their calls injected before/after the submodule forward call in the IR
- hooks/pre_hooks of the main module are still called from python but are ran as script functions
- hooks/pre_hooks added to concrete_type and class_type 
- compilation unit compiles hooks (stores them as methods, hooks can't have the same name as a method on a module)

Work left to do:
- having hooks accept default arguments if forward has default arguments (untested)
- writing checks and error messages for wrong hook/pre_hook function signatures (this is a huge need, currently wrong signatures will compile and then have runtime errors or will segfault)
- debug seg fault on jit's legacy fuser

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation"


Implementation of design laid out in: https://fb.quip.com/MY9gAqlroo0Z

Major changes:
- hooks/pre_hooks are compiled like methods on a module but are called with CallFunction
- hooks/pre_hooks of submodules have their calls injected before/after the submodule forward call in the IR
- hooks/pre_hooks of the main module are still called from python but are ran as script functions
- hooks/pre_hooks added to concrete_type and class_type 
- compilation unit compiles hooks (stores them as methods, hooks can't have the same name as a method on a module)

Work left to do:
- having hooks accept default arguments if forward has default arguments (untested)
- writing checks and error messages for wrong hook/pre_hook function signatures (this is a huge need, currently wrong signatures will compile and then have runtime errors or will segfault)
- debug seg fault on jit's legacy fuser

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation"


Implementation of design laid out in: https://fb.quip.com/MY9gAqlroo0Z

Major changes:
- hooks/pre_hooks are compiled like methods on a module but are called with CallFunction
- hooks/pre_hooks of submodules have their calls injected before/after the submodule forward call in the IR
- hooks/pre_hooks of the main module are still called from python but are ran as script functions
- hooks/pre_hooks added to concrete_type and class_type 
- compilation unit compiles hooks (stores them as methods, hooks can't have the same name as a method on a module)

Work left to do:
- having hooks accept default arguments if forward has default arguments (untested)
- writing checks and error messages for wrong hook/pre_hook function signatures (this is a huge need, currently wrong signatures will compile and then have runtime errors or will segfault)
- debug seg fault on jit's legacy fuser

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation"


Implementation of design laid out in: https://fb.quip.com/MY9gAqlroo0Z

Major changes:
- hooks/pre_hooks are compiled like methods on a module but are called with CallFunction
- hooks/pre_hooks of submodules have their calls injected before/after the submodule forward call in the IR
- hooks/pre_hooks of the main module are still called from python but are ran as script functions
- hooks/pre_hooks added to concrete_type and class_type 
- compilation unit compiles hooks (stores them as methods, hooks can't have the same name as a method on a module)

Work left to do:
- having hooks accept default arguments if forward has default arguments (untested)
- debug seg fault on jit's legacy fuser (see python testing commit)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation"


Implementation of design laid out in: https://fb.quip.com/MY9gAqlroo0Z

Major changes:
- hooks/pre_hooks are compiled like methods on a module but are called with CallFunction
- hooks/pre_hooks of submodules have their calls injected before/after the submodule forward call in the IR
- hooks/pre_hooks of the main module are still called from python but are ran as script functions
- hooks/pre_hooks added to concrete_type and class_type 
- compilation unit compiles hooks (stores them as methods, hooks can't have the same name as a method on a module)

Work left to do:
- having hooks accept default arguments if forward has default arguments (untested)
- debug seg fault on jit's legacy fuser (see python testing commit)

Differential Revision: [D25771122](https://our.internmc.facebook.com/intern/diff/D25771122)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation"


Implementation of design laid out in: https://fb.quip.com/MY9gAqlroo0Z

Major changes:
- hooks/pre_hooks are compiled like methods on a module but are called with CallFunction
- hooks/pre_hooks of submodules have their calls injected before/after the submodule forward call in the IR
- hooks/pre_hooks of the main module are still called from python but are ran as script functions
- hooks/pre_hooks added to concrete_type and class_type 
- compilation unit compiles hooks (stores them as methods, hooks can't have the same name as a method on a module)

Differential Revision: [D25771122](https://our.internmc.facebook.com/intern/diff/D25771122)

[ghstack-poisoned]

SplitInfinity commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -1681,6 +1704,43 @@ void initJitScriptBindings(PyObject* module) {
+          [](std::shared_ptr<ConcreteModuleType> concreteType,
+             const std::vector<Def>& hookDefs,
+             const std::vector<ResolutionCallback>& hookRcbs,
+             const std::vector<FunctionDefaults>& hookDefaults,
```
Does it make sense for hooks to have default values for arguments? Also, this doesn't seem to be used anywhere?

Lilyjjo replied:
Sorry, I forgot I had this logic here. I initially thought eager handled default arguments but it doesn't.

SplitInfinity commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -1681,6 +1704,43 @@ void initJitScriptBindings(PyObject* module) {
+             const std::vector<FunctionDefaults>& hookDefaults,
+             const std::vector<Def>& preHookDefs,
+             const std::vector<ResolutionCallback>& preHookRcbs,
+             const std::vector<FunctionDefaults>& preHookDefaults) {
```
Same here.

SplitInfinity commented on torch/jit/_recursive.py:
```diff
@@ -428,9 +447,20 @@ def init_fn(script_module):
     if concrete_type not in concrete_type_store.methods_compiled:
         create_methods_and_properties_from_stubs(concrete_type, method_stubs, property_stubs)
+        # Create hooks after methods to ensure no name collisions between hooks and methods.
+        # If done before hooks can overshadow methods that aren't exported
```
```suggestion
        # If done before, hooks can overshadow methods that aren't exported.
```

SplitInfinity commented on torch/csrc/jit/python/pybind_utils.cpp:
```diff
@@ -52,16 +52,29 @@ IValue toIValue(py::handle obj, const TypePtr& type, c10::optional<int32_t> N) {
-            py::str(obj),
-            " had a different number of elements than type ",
-            type->repr_str()));
+        if (!(tuple_size == 0 && type->repr_str() == "Tuple[None]")) {
```
Checking equality of `repr_str` is brittle - I suggest creating a `TupleType` instance with `NoneType` as the only element and using `TupleType::operator==` instead.

SplitInfinity commented on torch/csrc/jit/python/pybind_utils.cpp:
```diff
@@ -52,16 +52,29 @@ IValue toIValue(py::handle obj, const TypePtr& type, c10::optional<int32_t> N) {
-      values.reserve(tuple_size);
-      for (size_t i = 0; i < tuple_size; ++i) {
-        values.push_back(toIValue(tuple[i], elem_types[i]));
+      if (tuple_size == 0 && type->repr_str() == "Tuple[None]") {
```
Same here.

SplitInfinity reviewed (commented):
Nice! It seems that the relatively important issues I raised in my previous review have been addressed. Most of the issues I commented on this time are cosmetic. Let's discuss the changes to `pybind_utils.cpp` further offline to see if there is a more general improvement we can make.

Lilyjjo commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -4226,6 +4234,94 @@ std::vector<Function*> CompilationUnit::define(
+    functions.back()->ensure_defined();
+  }
+
+  return functions;
```
No it's not, I'll remove it.

Lilyjjo commented on torch/csrc/jit/python/python_sugared_value.h:
```diff
@@ -186,15 +186,110 @@ struct VISIBILITY_HIDDEN ModuleValue : public SugaredValue {
+      }
+      // create fake none if forward has no input
+      if (args.size() == 0) {
+        arg_values.push_back(caller.graph()->createNone()->output());
```
I think this one `None` value is already used by multiple hooks (I might be misunderstanding what you're trying to get at)

Lilyjjo commented on torch/csrc/jit/python/python_sugared_value.h:
```diff
@@ -186,15 +186,110 @@ struct VISIBILITY_HIDDEN ModuleValue : public SugaredValue {
+                    kwargs,
+                    n_binders)
+                ->asValue(loc, caller);
+        if (pre_hook_output->node()->output(0)->type() != NoneType::get()) {
+          if (pre_hook_output->node()->output(0)->type()->kind() !=
```
Oh wow they are the same, that's great. I didn't have the best mental model of how nodes and values related to each other when I was writing this ...

Commit:
Update on "[WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation"


Implementation of design laid out in: https://fb.quip.com/MY9gAqlroo0Z

Major changes:
- hooks/pre_hooks are compiled like methods on a module but are called with CallFunction
- hooks/pre_hooks of submodules have their calls injected before/after the submodule forward call in the IR
- hooks/pre_hooks of the main module are still called from python but are ran as script functions
- hooks/pre_hooks added to concrete_type and class_type 
- compilation unit compiles hooks (stores them as methods, hooks can't have the same name as a method on a module)

Differential Revision: [D25771122](https://our.internmc.facebook.com/intern/diff/D25771122)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation"


Implementation of design laid out in: https://fb.quip.com/MY9gAqlroo0Z

Major changes:
- hooks/pre_hooks are compiled like methods on a module but are called with CallFunction
- hooks/pre_hooks of submodules have their calls injected before/after the submodule forward call in the IR
- hooks/pre_hooks of the main module are still called from python but are ran as script functions
- hooks/pre_hooks added to concrete_type and class_type 
- compilation unit compiles hooks (stores them as methods, hooks can't have the same name as a method on a module)

Differential Revision: [D25771122](https://our.internmc.facebook.com/intern/diff/D25771122)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation"


Implementation of design laid out in: https://fb.quip.com/MY9gAqlroo0Z

Major changes:
- hooks/pre_hooks are compiled like methods on a module but are called with CallFunction
- hooks/pre_hooks of submodules have their calls injected before/after the submodule forward call in the IR
- hooks/pre_hooks of the main module are still called from python but are ran as script functions
- hooks/pre_hooks added to concrete_type and class_type 
- compilation unit compiles hooks (stores them as methods, hooks can't have the same name as a method on a module)

Differential Revision: [D25771122](https://our.internmc.facebook.com/intern/diff/D25771122)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation"


Implementation of design laid out in: https://fb.quip.com/MY9gAqlroo0Z

Major changes:
- hooks/pre_hooks are compiled like methods on a module but are called with CallFunction
- hooks/pre_hooks of submodules have their calls injected before/after the submodule forward call in the IR
- hooks/pre_hooks of the main module are still called from python but are ran as script functions
- hooks/pre_hooks added to concrete_type and class_type 
- compilation unit compiles hooks (stores them as methods, hooks can't have the same name as a method on a module)

Differential Revision: [D25771122](https://our.internmc.facebook.com/intern/diff/D25771122)

[ghstack-poisoned]

SplitInfinity commented on torch/csrc/jit/python/python_sugared_value.h:
```diff
@@ -186,15 +186,100 @@ struct VISIBILITY_HIDDEN ModuleValue : public SugaredValue {
 
-  // call module.forward
+  // call module.forward with pre_hooks and hooks
   std::shared_ptr<SugaredValue> call(
```
This method is no longer a one-liner. Can you move it into the .cpp file?

Lilyjjo commented on torch/csrc/jit/python/pybind_utils.cpp:
```diff
@@ -51,17 +51,30 @@ IValue toIValue(py::handle obj, const TypePtr& type, c10::optional<int32_t> N) {
+        values.reserve(tuple_size);
+        for (size_t i = 0; i < tuple_size; ++i) {
+          values.push_back(toIValue(tuple[i], elem_types[i]));
+        }
```
Tuple[None] modification made here

Commit:
Update on "[WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation"


Implementation of design laid out in: https://fb.quip.com/MY9gAqlroo0Z

Major changes:
- hooks/pre_hooks are compiled like methods on a module but are called with CallFunction
- hooks/pre_hooks of submodules have their calls injected before/after the submodule forward call in the IR
- hooks/pre_hooks of the main module are still called from python but are ran as script functions
- hooks/pre_hooks added to concrete_type and class_type 
- compilation unit compiles hooks (stores them as methods, hooks can't have the same name as a method on a module)

Differential Revision: [D25771122](https://our.internmc.facebook.com/intern/diff/D25771122)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation"


Implementation of design laid out in: https://fb.quip.com/MY9gAqlroo0Z

Major changes:
- hooks/pre_hooks are compiled like methods on a module but are called with CallFunction
- hooks/pre_hooks of submodules have their calls injected before/after the submodule forward call in the IR
- hooks/pre_hooks of the main module are still called from python but are ran as script functions
- hooks/pre_hooks added to concrete_type and class_type 
- compilation unit compiles hooks (stores them as methods, hooks can't have the same name as a method on a module)

Differential Revision: [D25771122](https://our.internmc.facebook.com/intern/diff/D25771122)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation"


Implementation of design laid out in: https://fb.quip.com/MY9gAqlroo0Z

Major changes:
- hooks/pre_hooks are compiled like methods on a module but are called with CallFunction
- hooks/pre_hooks of submodules have their calls injected before/after the submodule forward call in the IR
- hooks/pre_hooks of the main module are still called from python but are ran as script functions
- hooks/pre_hooks added to concrete_type and class_type 
- compilation unit compiles hooks (stores them as methods, hooks can't have the same name as a method on a module)

Differential Revision: [D25771122](https://our.internmc.facebook.com/intern/diff/D25771122)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation"


Implementation of design laid out in: https://fb.quip.com/MY9gAqlroo0Z

Major changes:
- hooks/pre_hooks are compiled like methods on a module but are called with CallFunction
- hooks/pre_hooks of submodules have their calls injected before/after the submodule forward call in the IR
- hooks/pre_hooks of the main module are still called from python but are ran as script functions
- hooks/pre_hooks added to concrete_type and class_type 
- compilation unit compiles hooks (stores them as methods, hooks can't have the same name as a method on a module)

Differential Revision: [D25771122](https://our.internmc.facebook.com/intern/diff/D25771122)

[ghstack-poisoned]

SplitInfinity commented on torch/csrc/jit/api/module_save.cpp:
```diff
@@ -5,11 +5,30 @@ namespace torch {
 namespace jit {
 
 void Module::save(std::ostream& out, const ExtraFilesMap& extra_files) const {
+#ifdef FBCODE_CAFFE2
+  if (this->type()->getForwardHooks().size() > 0 || this->type()->getForwardPreHooks().size() > 0) {
+    throw std::runtime_error(
+          "Cannot save module '" + this->type()->name()->name() 
+          + "' because it has forward hooks or pre-hooks attached. " +
+          "Saving modules with hooks not supported in FBCODE yet. Please " +
+          "remove the hooks before scripting if you want to save this model.");
+
+  }
+#endif
```
I would have put this in the serialization PR because that's what it's related to but this entire stack is landing together so I think it should be okay.

SplitInfinity reviewed (commented):
Looks good! I think no further action is needed here, unless you really want to move the one serialization change into the serialization PR. I will approve after the CI results are available.

Commit:
Update on "[WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation"


Implementation of design laid out in: https://fb.quip.com/MY9gAqlroo0Z

Major changes:
- hooks/pre_hooks are compiled like methods on a module but are called with CallFunction
- hooks/pre_hooks of submodules have their calls injected before/after the submodule forward call in the IR
- hooks/pre_hooks of the main module are still called from python but are ran as script functions
- hooks/pre_hooks added to concrete_type and class_type 
- compilation unit compiles hooks (stores them as methods, hooks can't have the same name as a method on a module)

Differential Revision: [D25771122](https://our.internmc.facebook.com/intern/diff/D25771122)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation"


Implementation of design laid out in: https://fb.quip.com/MY9gAqlroo0Z

Major changes:
- hooks/pre_hooks are compiled like methods on a module but are called with CallFunction
- hooks/pre_hooks of submodules have their calls injected before/after the submodule forward call in the IR
- hooks/pre_hooks of the main module are still called from python but are ran as script functions
- hooks/pre_hooks added to concrete_type and class_type 
- compilation unit compiles hooks (stores them as methods, hooks can't have the same name as a method on a module)

Differential Revision: [D25771122](https://our.internmc.facebook.com/intern/diff/D25771122)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation"


Implementation of design laid out in: https://fb.quip.com/MY9gAqlroo0Z

Major changes:
- hooks/pre_hooks are compiled like methods on a module but are called with CallFunction
- hooks/pre_hooks of submodules have their calls injected before/after the submodule forward call in the IR
- hooks/pre_hooks of the main module are still called from python but are ran as script functions
- hooks/pre_hooks added to concrete_type and class_type 
- compilation unit compiles hooks (stores them as methods, hooks can't have the same name as a method on a module)

Differential Revision: [D25771122](https://our.internmc.facebook.com/intern/diff/D25771122)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation"


Added support for module level hooks to be statically scripted: https://fb.quip.com/0IdqABRaPsAZ 
Allows hooks added with register_forward_hook and register_forward_pre_hook to be scripted and ran in python/cpp.

Major implementation changes:
- hooks/pre_hooks are compiled like methods on a module but are called with CallFunction
- hooks/pre_hooks of submodules have their calls injected before/after the submodule forward call in the IR
- hooks/pre_hooks of the main module are still called from python but are ran as script functions
- hooks/pre_hooks added to concrete_type and class_type 
- compilation unit compiles hooks (stores them as methods, hooks can't have the same name as a method on a module)

Differential Revision: [D25771122](https://our.internmc.facebook.com/intern/diff/D25771122)

[ghstack-poisoned]

SplitInfinity commented on torch/csrc/jit/python/python_sugared_value.cpp:
```diff
@@ -650,6 +650,7 @@ std::shared_ptr<SugaredValue> ModuleValue::call(
   // call pre_hooks
   if (have_pre_hooks) {
     for (const auto& hook : class_type->getForwardPreHooks()) {
+      assert(forward_input != nullptr);
```
Consider using `TORCH_INTERNAL_ASSERT` to provide more information about this if it is false.

SplitInfinity replied:
I'm also curious why you added it.

Lilyjjo replied:
The internal FB linter asked for me to add it-- I think it's not necessary because it shouldn't be null with the logic above it. Are we allowed to ignore FB lint warnings?

SplitInfinity replied:
I'd prefer not to ignore it but I do think we should try to use `TORCH_INTERNAL_ASSERT` so that we can provide some context or information if the assert ever trips.

SplitInfinity reviewed (approved):
Just a question about the assert that was added.

Commit:
Update on "[WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation"


Added support for module level hooks to be statically scripted: https://fb.quip.com/0IdqABRaPsAZ 
Allows hooks added with register_forward_hook and register_forward_pre_hook to be scripted and ran in python/cpp.

Major implementation changes:
- hooks/pre_hooks are compiled like methods on a module but are called with CallFunction
- hooks/pre_hooks of submodules have their calls injected before/after the submodule forward call in the IR
- hooks/pre_hooks of the main module are still called from python but are ran as script functions
- hooks/pre_hooks added to concrete_type and class_type 
- compilation unit compiles hooks (stores them as methods, hooks can't have the same name as a method on a module)

Differential Revision: [D25771122](https://our.internmc.facebook.com/intern/diff/D25771122)

[ghstack-poisoned]

Lilyjjo commented:
The internal FB clang linter had a warning about it, do you know the policy around ignoring lint errors? The logic prevents the pointer from being null so I don't think the assert is necessary. I took it out, but the warning on the diff is still there

SplitInfinity commented:
> The internal FB clang linter had a warning about it, do you know the policy around ignoring lint errors? The logic prevents the pointer from being null so I don't think the assert is necessary. I took it out, but the warning on the diff is still there

Given that we don't promote compilation warnings to errors, I think the policy is pretty lax. I think we should leave it in and convert it to a `TORCH_INTERNAL_ASSERT` if possible, but as you pointed out, `Node::output()` will crash if it tries to return `nullptr`.

Commit:
Update on "[WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation"


Added support for module level hooks to be statically scripted: https://fb.quip.com/0IdqABRaPsAZ 
Allows hooks added with register_forward_hook and register_forward_pre_hook to be scripted and ran in python/cpp.

Major implementation changes:
- hooks/pre_hooks are compiled like methods on a module but are called with CallFunction
- hooks/pre_hooks of submodules have their calls injected before/after the submodule forward call in the IR
- hooks/pre_hooks of the main module are still called from python but are ran as script functions
- hooks/pre_hooks added to concrete_type and class_type 
- compilation unit compiles hooks (stores them as methods, hooks can't have the same name as a method on a module)

Differential Revision: [D25771122](https://our.internmc.facebook.com/intern/diff/D25771122)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation"


Added support for module level hooks to be statically scripted: https://fb.quip.com/0IdqABRaPsAZ 
Allows hooks added with register_forward_hook and register_forward_pre_hook to be scripted and ran in python/cpp.

Major implementation changes:
- hooks/pre_hooks are compiled like methods on a module but are called with CallFunction
- hooks/pre_hooks of submodules have their calls injected before/after the submodule forward call in the IR
- hooks/pre_hooks of the main module are still called from python but are ran as script functions
- hooks/pre_hooks added to concrete_type and class_type 
- compilation unit compiles hooks (stores them as methods, hooks can't have the same name as a method on a module)

Differential Revision: [D25771122](https://our.internmc.facebook.com/intern/diff/D25771122)

[ghstack-poisoned]

siahuat0727 commented:
Hi @Lilyjjo  @SplitInfinity, I met the same issue with https://discuss.pytorch.org/t/type-mismatch-with-hooks-input-when-using-jit/137783.
I would really appreciate it if you could please shed some light on this matter.
